packages/run-protocol/src/vaultFactory/liquidateIncrementally.js

```diff
@@ -0,0 +1,322 @@
+/* eslint-disable no-await-in-loop */
+// no @ts-check
+
+import { E } from '@endo/eventual-send';
+import {
+  getAmountOut,
+  makeRatio,
+  assertProposalShape,
+  offerTo,
+  natSafeMath as NatMath,
+} from '@agoric/zoe/src/contractSupport/index.js';
+import { AmountMath } from '@agoric/ertp';
+import { Far } from '@endo/marshal';
+
+import { ceilMultiplyBy } from '@agoric/zoe/src/contractSupport/ratio';
+import { makeTracer } from '../makeTracer.js';
+// import { quote } from '@agoric/assert';
+
+const { details: X } = assert;
+const trace = makeTracer('LiqI');
+
+/**
+ * This contract liquidates a payment in multiple tranches to limit the
+ * price impact an the AMM of any one sale. Each block it will compute
+ * a tranche of collateral to sell, where the size is a function of
+ * the amount of that collateral in the AMM pool and the desired price impact.
+ * It then gets 3 prices for the current tranche:
+ * - AMM quote - compute xYK locally based on the pool sizes
+ * - Reserve quote - based on a low price at which the Reserve will purchase
+ * - Oracle quote - provide by the oracle that triggered liquidation
+ * - Oracle limit - a limit that is DIVERGENCE_TOLERANCE below the
+ *   Oracle quote.
+ * These sell based on the following:
+ * - If the AMM quote is *below* the oracle limit, skip this block
+ *   to wait for oracle and AMM prices to converge.
+ *    - should it still sell to teh Reserve here?
+ * - If the Reserve quote is higher than the AMM quote, sell to the
+ *   Reserve with a `want` of the Oracle limit
+ * - Otherwise, sell to the AMM with a `want` of the Oracle limit
+ * If selling a tranche for the minimum would pay off the debt (`isLast`),
+ * then use sell using`swapOut` with a minimum of the debt. If that fails
+ * this is not the last tranche, tehn sell with the `swapIn` operation,
+ * using the oracle price reference as the minimum.
+ *
+ * TODO integrate the reserve
+ */
+
+/** @type {ContractStartFn} */
+const start = async zcf => {
+  const { amm, priceAuthority, timerService, debtBrand } =
+    /** @type {LiquidationContractTerms} */ zcf.getTerms();
+  const nextBlock = async () => E(timerService).delay(1n);
+
+  // TODO make subject to governance
+  const SCALE = 1_000_000n;
+  const BASIS_POINTS = 10_000n * SCALE;
+  const BP2 = BASIS_POINTS * BASIS_POINTS;
+
+  const AMM_FEE_BP = 30n * SCALE;
+  const MAX_IMPACT_BP = 50n * SCALE;
+  // sqrt(impact + 1) - 1
+  // Since we take SQRT but want to be BP, multiply in BP
+  const IMPACT_FACTOR_BP =
+    BigInt(
+      Math.ceil(
+        Math.sqrt(Number((MAX_IMPACT_BP + BASIS_POINTS) * BASIS_POINTS)),
+      ),
+    ) - BASIS_POINTS;
+  const IMPACT_FACTOR_BP2 = IMPACT_FACTOR_BP * (BASIS_POINTS - AMM_FEE_BP);
```

Comment:
since these will be subject to governance, they should be coded as functions of the values that will be governable. (we can't make `SCALE` subject to governance yet but we know that it won't be constant)

Reply:
These need to be ripped out and replaced with governance. That's the separate ticket for governance.